### PR TITLE
Create a basic NewEntry form

### PR DESCRIPTION
**Pull request by frozenhelium**

### CODE REVIEW

File: src/views/NewEntry/index.tsx [hunk 1]
```diff
@@ -191,9 +261,48 @@ interface NewEntryProps {
...
+    );
+    const { data } = useQuery(ENTRY_OPTIONS);
+    const [
+        countryOptions,
+        crisisTypeOptions,
+        crisisOptions,
+    ] = React.useMemo(() => ([
+        data?.countryList?.results || [],
+        data?.__type?.enumValues || [],
+        data?.crisisList?.results || [],
```
Comment (tnagorra) [inline src/views/NewEntry/index.tsx]: we should use `??` syntax to speficy default value.
We should not create empty list everytime (even if it is inside memo, they are bundled up

File: src/views/NewEntry/index.tsx [hunk 2]
```diff
@@ -191,9 +261,48 @@ interface NewEntryProps {
...
+    ]), [data]);
+
+    console.info(data);
```
Comment (tnagorra) [inline src/views/NewEntry/index.tsx]: Let's not leave console logs in render function

File: src/views/NewEntry/index.tsx [hunk 3]
```diff
@@ -191,9 +261,48 @@ interface NewEntryProps {
...
+            publisher,
+            publishDate,
+        } = finalValue.sourceDetails;
+
+        const entry = {
+            articleTitle,
+            source,
+            publisher,
+            publishDate,
+            event: 1,
```
Comment (tnagorra) [inline src/views/NewEntry/index.tsx]: I think we should at least mark that event is hard-coded here.
So, it we can later fix this.

File: src/views/NewEntry/StrataInput.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+interface StrataInputProps {
+    index: number;
+    value: StrataFormProps;
+    error: Error<StrataFormProps> | undefined;
+    onChange: (value: StrataFormProps, index: number) => void;
+    onRemove: (index: number) => void;
+    className?: string;
+}
+
+function StrataInput(p: StrataInputProps) {
```
Comment (tnagorra) [inline src/views/NewEntry/StrataInput.tsx:23]: You mean `props` ?

File: src/views/NewEntry/ReviewInput.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { _cs } from '@togglecorp/fujs';
+import {
+    TextInput,
+    TabPanel,
+} from '@togglecorp/toggle-ui';
+
+import styles from './styles.css';
+
+function Review() {
```
Comment (tnagorra) [inline src/views/NewEntry/ReviewInput.tsx:10]: Is this a WIP?

File: src/views/NewEntry/FigureInput.tsx [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+        termOptions,
+        roleOptions,
+        typeOptions,
+    ] = React.useMemo(() => [
+        data?.quantifierList?.enumValues,
+        data?.unitList?.enumValues,
+        data?.termList?.enumValues,
+        data?.roleList?.enumValues,
+        data?.typeList?.enumValues,
+    ], [data]);
```
Comment (tnagorra) [inline src/views/NewEntry/FigureInput.tsx:101]: I don't even understand why this is memoized

File: src/views/NewEntry/FigureInput.tsx [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+
+    const onValueChange = useFormObject<number, FigureFormProps>(index, value, onChange);
+
+    const handleAgeAdd = React.useCallback(() => {
+        const uuid = new Date().getTime();
```
Comment (tnagorra) [inline src/views/NewEntry/FigureInput.tsx:106]: Woah! Woah! uuid = timestamp alert!

File: src/views/NewEntry/FigureInput.tsx [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+            'strataJson',
+        );
+    }, [onValueChange, value]);
+
+    const {
+        onValueChange: onStrataChange,
+        onValueRemove: onStrataRemove,
+    } = useFormArray('strataJson', value.strataJson, onValueChange);
+
+    return (
```
Comment (tnagorra) [inline src/views/NewEntry/FigureInput.tsx:133]: Error for each input should be passed as `error` prop

File: src/views/NewEntry/DetailsInput.tsx [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    const [showAddEventModal, setShowAddEventModal] = React.useState(false);
+
+    const onValueChange = useFormObject<K, DetailsFormProps>(name, value, onChange);
+    const handleAddEventButtonClick = React.useCallback(() => {
+        setShowAddEventModal(true);
+    }, [setShowAddEventModal]);
+
+    const handleAddEventModalClose = React.useCallback(() => {
+        setShowAddEventModal(false);
+    }, [setShowAddEventModal]);
```
Comment (tnagorra) [inline src/views/NewEntry/DetailsInput.tsx:64]: Let's create a hook to handle these use cases

Reference: https://github.com/the-deep/client/blob/a15d2453e8565a20896992379e1cbdf05239d9fd/src/hooks/stateManagement.ts

File: src/views/NewEntry/DetailsInput.tsx [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+    const {
+        data,
+        refetch: refetchDetailOptions,
+    } = useQuery(DETAILS_OPTIONS);
+
+    const eventOptions = React.useMemo(() => (
+        data?.eventList?.results || []
+    ), [data]);
```
Comment (tnagorra) [inline src/views/NewEntry/DetailsInput.tsx:73]: Same as previous comment to use `??`

Comment (tnagorra) [reply src/views/NewEntry/DetailsInput.tsx:73]: IMO, we should let SelectInput handle these cases.

File: src/views/NewEntry/AgeInput.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+interface AgeInputProps {
+    index: number;
+    value: AgeFormProps;
+    error: Error<AgeFormProps> | undefined;
+    onChange: (value: AgeFormProps, index: number) => void;
+    onRemove: (index: number) => void;
+    className?: string;
+}
+
+function AgeInput(p: AgeInputProps) {
```
Comment (tnagorra) [inline src/views/NewEntry/AgeInput.tsx:23]: We prefer `props` instead of `p`.

Comment (tnagorra) [reply src/views/NewEntry/AgeInput.tsx:23]: @AdityaKhatri any thoughts?

File: src/components/CrisisForm/index.tsx [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    } = props;
+
+    const { data } = useQuery(CRISIS_OPTIONS);
+    const [
+        countryOptions,
+        crisisTypeOptions,
+    ] = React.useMemo(() => ([
+        data?.countryList?.results || [],
+        data?.__type?.enumValues || [],
+    ]), [data]);
```
Comment (tnagorra) [inline src/components/CrisisForm/index.tsx:99]: Same comments as previously stated.

File: src/components/EventForm/index.tsx [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+const defaultFormValues: EventFormFields = {
+    countries: [],
+    crisis: '',
+    eventType: '',
+    name: '',
+};
+
+const subTypesSelector = (d: { subTypes: unknown[] }) => d.subTypes;
+
+function EventForm(p: EventFormProps) {
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:150]: Did you mean `props`?

File: src/components/EventForm/index.tsx [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+    const {
+        value: initialFormValues = defaultFormValues,
+        onEventCreate,
+    } = p;
+
+    const [showCrisisFormModal, setShowCrisisFormModal] = React.useState(false);
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:156]: Let' use a hook for this common usecase

File: src/components/EventForm/index.tsx [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+        data?.crisisList?.results,
+        data?.disasterCategoryList,
+        data?.disasterSubCategoryList,
+        data?.disasterTypeList,
+        data?.triggerList,
+        data?.violenceList,
+        data?.__type?.enumValues || [],
+        listToMap(data?.violenceList || [], basicEntityKeySelector, subTypesSelector),
+        listToMap(data?.triggerList || [], basicEntityKeySelector, subTypesSelector),
+    ]), [data]);
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:203]: We should handle these individually.
Most of them don't event need memoization.
Also, prefer `??` syntax to `||` for default values.

File: src/components/PageHeader/index.tsx [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        title,
+        actions,
+    } = props;
+
+    return (
+        <header className={_cs(styles.header, className)}>
+            <div className={styles.headingSection}>
+                <h1>
+                    { title }
+                </h1>
```
Comment (tnagorra) [inline src/components/PageHeader/index.tsx:24]: I think we should have a "tooltip" on hover and then have ellipses if needs to break.

File: src/views/NewEntry/DetailsInput.tsx (hunk 9 above)
Comment (frozenhelium) [reply src/views/NewEntry/DetailsInput.tsx:63]: The modal might get removed later

File: src/components/EventForm/index.tsx (hunk 14 above)
Comment (frozenhelium) [reply src/components/EventForm/index.tsx:157]: might get removed later

File: src/components/PageHeader/index.tsx (hunk 16 above)
Comment (frozenhelium) [reply src/components/PageHeader/index.tsx:24]: Can be added as we go

File: src/components/CrisisForm/index.tsx [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+    } = props;
+
+    const { data } = useQuery(CRISIS_OPTIONS);
+    const [
+        countryOptions,
+        crisisTypeOptions,
+    ] = React.useMemo(() => ([
+        data?.countryList?.results ?? emptyList,
+        data?.__type?.enumValues ?? emptyList,
+    ]), [data]);
```
Comment (tnagorra) [inline src/components/CrisisForm/index.tsx:101]: We don't need to memorize this at all;

File: src/components/EventForm/index.tsx [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+        data?.crisisList?.results ?? emptyList,
+        data?.disasterCategoryList ?? emptyList,
+        data?.disasterSubCategoryList ?? emptyList,
+        data?.disasterTypeList ?? emptyList,
+        data?.triggerList ?? emptyList,
+        data?.violenceList ?? emptyList,
+        data?.__type?.enumValues ?? emptyList,
+        listToMap(data?.violenceList ?? emptyList, basicEntityKeySelector, subTypesSelector),
+        listToMap(data?.triggerList ?? emptyList, basicEntityKeySelector, subTypesSelector),
+    ]), [data]);
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:204]: We don't need to memoize this

File: src/components/EventForm/index.tsx [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+                        keySelector={basicEntityKeySelector}
+                        labelSelector={basicEntityLabelSelector}
+                        label="Trigger"
+                        name="trigger"
+                        value={value.trigger}
+                        onChange={onValueChange}
+                        error={error?.fields?.trigger}
+                    />
+                    <SelectInput
+                        options={triggerSubTypeOptions[value.trigger] || []}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:300]: use `??` syntax

File: src/components/EventForm/index.tsx [hunk 20]
```diff
@@ -0,0 +1,389 @@
...
+                        options={violenceOptions}
+                        keySelector={basicEntityKeySelector}
+                        labelSelector={basicEntityLabelSelector}
+                        label="Type of Violence"
+                        name="violence"
+                        value={value.violence}
+                        onChange={onValueChange}
+                    />
+                    <SelectInput
+                        options={violenceSubTypeOptions[value.violence] || []}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:320]: Prefer `??`

File: src/views/NewEntry/DetailsInput.tsx (hunk 9 above)
Comment (tnagorra) [reply src/views/NewEntry/DetailsInput.tsx:63]: If you do need this, it's good if you use a common hook.
Will close this for now!

File: src/views/NewEntry/DetailsInput.tsx [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+    }, [setShowAddEventModal]);
+
+    const {
+        data,
+        refetch: refetchDetailOptions,
+    } = useQuery(DETAILS_OPTIONS);
+
+    const eventOptions = React.useMemo(() => (
+        data?.eventList?.results ?? emptyList
+    ), [data]);
```
Comment (tnagorra) [inline src/views/NewEntry/DetailsInput.tsx:72]: No need to memoize this

File: src/views/NewEntry/FigureInput.tsx [hunk 22]
```diff
@@ -0,0 +1,429 @@
...
+        termOptions,
+        roleOptions,
+        typeOptions,
+    ] = React.useMemo(() => [
+        data?.quantifierList?.enumValues ?? emptyList,
+        data?.unitList?.enumValues ?? emptyList,
+        data?.termList?.enumValues ?? emptyList,
+        data?.roleList?.enumValues ?? emptyList,
+        data?.typeList?.enumValues ?? emptyList,
+    ], [data]);
```
Comment (tnagorra) [inline src/views/NewEntry/FigureInput.tsx:104]: no need to memoize this

File: src/views/NewEntry/index.tsx [hunk 23]
```diff
@@ -1,285 +1,367 @@
...
+                        return {
+                            ...basicFields,
+                            ...disaggregatedFields,
+                        };
                     }
-                    return fields;
+
+                    return {
+                        ...basicFields,
+                    };
```
Comment (tnagorra) [inline src/views/NewEntry/index.tsx:151]: Why not?
```suggestion
                    return basicFields;
```

File: src/views/NewEntry/ReviewInput.tsx (hunk 5 above)
Comment (frozenhelium) [reply src/views/NewEntry/ReviewInput.tsx:10]: yes

File: src/views/NewEntry/DetailsInput.tsx [hunk 24]
```diff
@@ -84,7 +93,7 @@ function DetailsInput<K extends string>(props: DetailsInputProps<K>) {
                     labelSelector={basicEntityLabelSelector}
                     name="event"
                     onChange={onValueChange}
-                    options={data?.eventList?.results ?? emptyList}
+                    options={data?.eventList?.results ?? emptyList as BasicEntity[]}
```
Comment (tnagorra) [inline src/views/NewEntry/DetailsInput.tsx:96]: No need to pass emptyList to options

File: src/components/EventForm/index.tsx [hunk 25]
```diff
@@ -181,8 +199,16 @@ function EventForm(props: EventFormProps) {
...
+            data?.violenceList ?? emptyBasicEntityWithSubTypesList,
+            basicEntityKeySelector,
+            subTypesSelector,
+        ),
+        listToMap(
+            data?.triggerList ?? emptyBasicEntityWithSubTypesList,
+            basicEntityKeySelector,
+            subTypesSelector,
+        ),
     ]), [data]);
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx]: Use separate useMemo

File: src/components/EventForm/index.tsx [hunk 26]
```diff
@@ -222,7 +242,7 @@ function EventForm(props: EventFormProps) {
             >
                 <div className={styles.crisisRow}>
                     <SelectInput
-                        options={data?.crisisList?.results ?? emptyList}
+                        options={data?.crisisList?.results ?? emptyBasicEntityList}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:245]: Why do this? we don't need to send emptylist on selectinput now

File: src/components/EventForm/index.tsx [hunk 27]
```diff
@@ -250,7 +281,7 @@ function EventForm(props: EventFormProps) {
                 </div>
                 <div className={styles.twoColumnRow}>
                     <SelectInput
-                        options={data?.eventType?.enumValues ?? emptyList}
+                        options={data?.eventType?.enumValues ?? emptyEnumEntityList}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:284]: why do we need to send default value for options here?
we can even have a work around and have a default empty list alue inside the component itself

File: src/components/EventForm/index.tsx [hunk 28]
```diff
@@ -267,59 +298,84 @@ function EventForm(props: EventFormProps) {
...
+                                name="violence"
+                                value={value.violence}
+                                onChange={onValueChange}
+                            />
+                            <SelectInput
+                                options={value.violence ? (
+                                    violenceSubTypeOptions[value.violence] ?? emptyBasicEntityList
+                                ) : (
+                                    emptyBasicEntityList
+                                )}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:343]: same comment regarding empty list on select input

File: src/components/EventForm/index.tsx [hunk 29]
```diff
@@ -267,59 +298,84 @@ function EventForm(props: EventFormProps) {
...
+                                labelSelector={basicEntityLabelSelector}
+                                label="Sub-type"
+                                name="triggerSubType"
+                                value={value.triggerSubType}
+                                onChange={onValueChange}
+                            />
+                        </div>
+                        <div className={styles.twoColumnRow}>
+                            <SelectInput
+                                options={data?.violenceList ?? emptyBasicEntityList}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:330]: same comment regarding empty list on select input

File: src/components/EventForm/index.tsx [hunk 30]
```diff
@@ -267,59 +298,84 @@ function EventForm(props: EventFormProps) {
...
+                                name="trigger"
+                                value={value.trigger}
+                                onChange={onValueChange}
+                                error={error?.fields?.trigger}
+                            />
+                            <SelectInput
+                                options={value.trigger ? (
+                                    triggerSubTypeOptions[value.trigger] ?? emptyBasicEntityList
+                                ) : (
+                                    emptyBasicEntityList
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:318]: same comment regarding empty list on select input

File: src/components/EventForm/index.tsx [hunk 31]
```diff
@@ -267,59 +298,84 @@ function EventForm(props: EventFormProps) {
                         error={error?.fields?.glideNumber}
                     />
                 </div>
+                { value.eventType === 'CONFLICT' && (
+                    <>
+                        <div className={styles.twoColumnRow}>
+                            <SelectInput
+                                options={data?.triggerList ?? emptyBasicEntityList}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:305]: same comment regarding empty list on select input

File: src/components/EventForm/index.tsx [hunk 32]
```diff
@@ -267,59 +298,84 @@ function EventForm(props: EventFormProps) {
...
-                        keySelector={basicEntityKeySelector}
-                        labelSelector={basicEntityLabelSelector}
-                        label="Actor"
-                        name="actor"
-                        value={value.actor}
-                        onChange={onValueChange}
-                    />
+                    { value.eventType === 'DISASTER' && (
+                        <SelectInput
+                            options={data?.disasterSubTypeList ?? emptyBasicEntityList}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:357]: same comment regarding empty list on select input

File: src/components/EventForm/index.tsx [hunk 33]
```diff
@@ -267,59 +298,84 @@ function EventForm(props: EventFormProps) {
...
+                            labelSelector={basicEntityLabelSelector}
+                            label="Disaster type"
+                            name="disasterSubType"
+                            value={value.disasterSubType}
+                            onChange={onValueChange}
+                        />
+                    )}
+                    { value.eventType === 'CONFLICT' && (
+                        <SelectInput
+                            options={data?.actorList ?? emptyBasicEntityList}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:368]: v

File: src/components/EventForm/index.tsx [hunk 34]
```diff
@@ -267,59 +298,84 @@ function EventForm(props: EventFormProps) {
...
+                            labelSelector={basicEntityLabelSelector}
+                            label="Actor"
+                            name="actor"
+                            value={value.actor}
+                            onChange={onValueChange}
+                        />
+                    )}
                     <MultiSelectInput
-                        options={data?.countryList?.results ?? emptyList}
+                        options={data?.countryList?.results ?? emptyBasicEntityList}
```
Comment (tnagorra) [inline src/components/EventForm/index.tsx:378]: same comment regarding empty list on select input

File: src/types/index.ts [hunk 35]
```diff
@@ -4,10 +4,18 @@ export interface BasicEntity {
 }
 
 export interface EnumEntity {
-    name: string;
+    name: string & number;
```
Comment (tnagorra) [inline src/types/index.ts:7]: did you mean `string | number` ?

File: src/types/index.ts [hunk 36]
```diff
@@ -123,15 +130,15 @@ export interface FigureFormProps {
     isDisaggregated: boolean;
     locationCamp?: number;
     locationNonCamp?: number;
-    quantifier?: number;
+    quantifier: number & undefined;
```
Comment (tnagorra) [inline src/types/index.ts:133]: What does `& <literal>` mean?

File: src/views/NewEntry/FigureInput.tsx [hunk 37]
```diff
@@ -109,7 +114,7 @@ function FigureInput(props: FigureInputProps) {
         const uuid = uuidv4();
         const newStrata = { uuid };
         onValueChange(
-            [...value.strataJson, newStrata],
+            [...value.strataJson, newStrata] as StrataFormProps[],
```
Comment (tnagorra) [inline src/views/NewEntry/FigureInput.tsx:117]: Why cast here?

**Comment by tnagorra** [comment]
Currently, the error handling for NewEvent modal is problematic.
We should look also check if it is the case for other forms too.

**Comment by tnagorra** [comment]
Did we find out why all the forms are submitting?

**Pull request merged**